index.bs

```diff
@@ -2089,6 +2109,39 @@ To <dfn>await a navigation</dfn> given |context|, |request|, |wait condition|, a
 
 </div>
 
+#### The browsingContext.Locator Type #### {#type-browsingContext-Locator}
+
+[=remote end definition=] and [=local end definition=]
+
+<pre class="cddl remote-cddl local-cddl">
+browsingContext.Locator = (
+   browsingContext.CssLocator /
+   browsingContext.TextLocator /
+   browsingContext.XPathLocator
+)
+
+browsingContext.CssLocator = {
+   type: "css",
+   value: text
+}
+
+browsingContext.TextLocator {
```

Comment:
I'm slightly against the idea of having a text selector with WebDriver BiDi. The reasons are:

 - There is no support of ShadowDOM,
 - Text could also include `input` values and other properties,
 - It's almost guaranteed to be less capable than a well-written text searcher built client-side.

To stay compatible with WebDriver, I'd be in favor of just adding the `link text` and `partial link text` algorithm verbatim here. On the other hand, we could also update the algorithm to be actually useful (i.e. fixing the above problems), however this is itself a large problem.

All in all, perhaps we can split this PR into two PRs, one for text locators and one for the rest.

Reply:
Splitting up sounds great. This would allow us to start sooner than later on the implementation.

Reply:
> There is no support of ShadowDOM,

That's currently true, in the sense that one doesn't get the text from within shadow DOM. But it's also true that it's impossible for the js-based solutions to work with closed shadow DOM, whereas a native implementation could do. 

In general, better support for shadow DOM seems like a good reason to build in element selection rather than depending only on js, rather than the other way around.

> Text could also include input values and other properties,

I can't really tell what you mean here. `innerText` doesn't contain the input value. 

> It's almost guaranteed to be less capable than a well-written text searcher built client-side.

I don't think this is true. A text-searcher written client side can't pierce closed shadow DOM without making multiple roundtrips, and getting the same layout information as `innerText` uses is challenging. On the other hand depending specifically on `innerText` does sacrifice quite a lot of flexibility and some performance (assuming one doesn't have the ability to cache partial innerText values).

So it seems like engineering tradeoffs.

Reply:
> I don't think this is true.
 
I made an assumption that a theoretical text searcher in WebDriver BiDi would not be descriptive enough to allow handling this case. Given that querying closed shadow roots are not part of any spec, I assume there are many obstacles with writing a text searcher that can fix all details.

> I can't really tell what you mean here.

In Puppeteer and Playwright, `input.value` is considered to be the "innerText" of an input element (since input elements don't have innerText).

Reply:
> It's almost guaranteed to be less capable than a well-written text searcher built client-side.

Who is going to write it? I don't think it will get written when it is really useful for the community to have from the start.

Users will Hightlight text and then paste that into their test and say "find this text". This is how a lot of users have used Find By Link Text in webdriver classic. 

> Text could also include input values and other properties,
Users already know that to get the value of the input they need to get it via a property or attributes. I can't see them looking for elements based on text that way. 

The use of innerText here would would be better. Browser vendors can spend time improving the algorithm to make it less expensive. The Selenium code around visibility doesn't do any form of caching and it's never been too bad expect on text heavy websites like the HTML spec.

Reply:
> Who is going to write it? I don't think it will get written when it is really useful for the community to have from the start.

There was an implication that it's already been done; e.g. Puppeteer and Playwright.

Reply:
@jrandolf It's been repeatedly made clear that Puppeteer has no intention of using this API (unless something has changed to make that statement obsolete), and Playwright has expressed an explicit disinterest in supporting use cases that this would enable (e.g., non-Web or mobile use cases). I've already culled a great deal from the initial proposal that would make things far more useful (explicit ShadowRoot and Frame locators, batching of location steps to minimize round trips, and so on). This is a proverbial hill I'm willing to make a stand on.

I do not want to artificially cripple a "find by visible text" locator (e.g., by making it only applicable to `<a>` elements), so as to allow frameworks based on WebDriver BiDi a chance to compete with frameworks based on other, non-open, proprietary technologies. The fact that those other frameworks may consider the `value` property of elements that do not produce [CSS boxes when rendered](https://html.spec.whatwg.org/multipage/dom.html#rendered-text-collection-steps) as part of the "innerText" of the element is immaterial to me. This is especially so since (1) that's not relying on standardized behavior, and (2) those projects are unlikely to use this mechanism in the first place, by their own admission.

Reply:
There were 2 main value propositions here as I see it.

Firstly, remove the need to inject JS over the wire to do simple things. Time and again we've had members going "we'll just inject javaScript" which is over time is going impact the performance of people using cloud services. If we inject things on session create, or in the mobile case, we going negatively impact the way webdriver works. WebDriver Classic works really well on mobile devices (except for native mobile dialogs). The non-web case also benefits from this as they can then figure out what to do find elements without having to break down the JS and guess where the user wants the element from.

Secondly, it was a way to give a meaningful way to find the elements with minimal round trips. Again this benefits cloud providers immensely, including Playwright as they branch out into the cloud space, hopefully helping people run their tests faster. It normalises what tools like NightwatchJS, Playwright, Puppeteer are doing with a lot of round trips and tries to minimise that. Other JS frameworks like Testing Library would benefit from this. The outlier is going to be closed shadow roots (and that should have been solved by the web components group ages ago but that's a different argument)

Reply:
I think what would satisfy me is just renaming this to `inner-text` rather than `text`.

Reply:
@jrandolf So renamed in latest commit.